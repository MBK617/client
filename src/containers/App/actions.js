export const setToken = (token) => ({ type: 'SET_TOKEN', token });
export const setTokenSuccess = (token) => ({ type: 'SET_TOKEN_SUCCESS', token });
export const getToken = () => ({ type: 'GET_TOKEN' });
export const getTokenSuccess = (token) => ({ type: 'GET_TOKEN_SUCCESS', token });

export const pushHistory = (path) => ({ type: 'PUSH_HISTORY', path });
export const popHistory = () => ({ type: 'POP_HISTORY' });