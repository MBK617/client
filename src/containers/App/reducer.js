/*
 * AppReducer
 */

import { fromJS } from 'immutable';

export const initialState = fromJS({
  token: null,
}).set('history', ['home']);

function appReducer(state = initialState, action) {
  switch (action.type) {
    case 'GET_TOKEN_SUCCESS':
      return state.set('token', action.token);
    case 'SET_TOKEN_SUCCESS':
      return state.set('token', action.token);
    case 'PUSH_HISTORY':
      return state.set('history', [action.path, ...state.get('history')]);
    case 'POP_HISTORY':
      return state.set('history', state.get('history').slice(1));
    default:
      return state;
  }
}

export default appReducer;