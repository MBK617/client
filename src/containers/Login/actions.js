export const logIn = (email, password) => ({ type: 'LOGIN', email, password })
export const logInSuccess = (token) => ({ type: 'LOGIN_SUCCESS', token })
export const logInFailure = (error) => ({ type: 'LOGIN_FAILURE', error })