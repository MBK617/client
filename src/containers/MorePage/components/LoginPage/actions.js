export const logIn = (email, password) => ({ type: 'LOGIN', email, password })
export const logInSuccess = () => ({ type: 'LOGIN_SUCCESS' })
export const logInFailure = (error) => ({ type: 'LOGIN_FAILURE', error })