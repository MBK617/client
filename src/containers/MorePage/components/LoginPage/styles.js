import { StyleSheet } from 'react-native'

const styles = StyleSheet.create({
  root: {
    height: '100%',
    padding: 40,
    justifyContent: 'center',
  },
  error: {
    textAlign: 'center',
    color: '#ff0000',
    width: 250,
    marginLeft: 'auto',
    marginRight: 'auto',
    marginBottom: 10,
  }
});

export default styles;