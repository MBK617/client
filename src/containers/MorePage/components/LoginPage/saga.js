import { put, takeLatest, all, call } from 'redux-saga/effects';
import { getRequestOptions, request } from "utils/request";
import { setToken } from "containers/App/actions";
import { logInSuccess, logInFailure } from "./actions";

export function* makeLoginRequest(action) {
  const payload = {
    userEmail: action.email,
    userPassword: action.password
  }
  try {
    const result = yield call(request, `/api/public/user/login`, getRequestOptions('post', payload));
    yield put(setToken(result.token));
    yield put(logInSuccess());
  } catch (err) {
    yield put(logInFailure(err.message));
  }
}

export default function* rootSaga() {
  yield all([
    takeLatest('LOGIN', makeLoginRequest),
  ]);
}