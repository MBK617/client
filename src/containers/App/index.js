import React from 'react';
import { View } from 'react-native';

import Counter from 'containers/Counter';

import styles from './styles';

function App() {
  return (
    <View style={styles.container}>
      <Counter/>
    </View>
  );
}

export default App;