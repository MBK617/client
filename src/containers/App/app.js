import React, { useEffect } from 'react';
import { View, Image } from 'react-native';
import { compose } from 'redux';

import { 
  Appbar, 
  BottomNavigation 
} from 'react-native-paper';

import injectSaga from 'utils/redux/injectSaga';
import injectReducer from 'utils/redux/injectReducer';

import useHistory from 'hooks/useHistory';
import useLoginStatus from 'hooks/useLoginStatus';

import styles from './styles';
import reducer from './reducer';
import saga from './saga';

import HomePage from '../HomePage';
import EventsPage from '../EventsPage';
import InboxPage from '../InboxPage';
import MorePage from '../MorePage';

const pages = [
  { key: 'home', icon: 'home', authenticated: false },
  { key: 'events', icon: 'calendar', authenticated: true },
  { key: 'inbox', icon: 'forum', authenticated: true },
  { key: 'more', icon: 'dots-horizontal', authenticated: false },
]

const App = () => {
  const isLoggedIn = useLoginStatus();
  const { history, path, goTo, goBack } = useHistory();

  const [index, setIndex] = React.useState(0);
  const [routes, setRoutes] = React.useState(pages.filter(({authenticated}) => isLoggedIn || !authenticated).map(({ authenticated, ...page }) => page));

  useEffect(() => {
    const newRoutes = pages.filter(({authenticated}) => isLoggedIn || !authenticated).map(({ authenticated, ...page }) => page);
    setIndex(newRoutes.findIndex(route => route.key === routes[index].key))
    setRoutes(newRoutes);
  }, [isLoggedIn])

  useEffect(() => {
    const newPath = routes[index].key;
    goTo(newPath);
  }, [index])

  useEffect(() => {
    const newIndex = routes.findIndex(page => page.key === path?.split('/')[0]);
    setIndex(newIndex);
  }, [path])

  const renderScene =  () => {
    switch (path?.split('/')[0]) {
      case 'home':
        return <HomePage />;
      case 'events':
        return <EventsPage />;
      case 'inbox':
        return <InboxPage />;
      case 'more':
        return <MorePage />;
    }
  }
  
  return (
    <View style={styles.root}>
      <Appbar.Header 
        statusBarHeight={0}
        style={styles.header}
      > 
        <Appbar.BackAction onPress={goBack} style={{opacity: history.length > 1 ? 1 : 0, transition: '.2s'}}/>
        <Image source={require('assets/mbklogo.jpg')} style={styles.logo}/>
        <Appbar.Action icon="bell" />
      </Appbar.Header>
      <BottomNavigation
        barStyle={{ backgroundColor: '#222' }}
        navigationState={{ index, routes }}
        onIndexChange={setIndex}
        renderScene={renderScene}
        shifting={false} 
        labeled={false} 
      />
    </View>
  );
}

const withReducer = injectReducer({ key: 'app', reducer });
const withSaga = injectSaga({ key: 'app', saga });

export default compose(withReducer, withSaga)(App);