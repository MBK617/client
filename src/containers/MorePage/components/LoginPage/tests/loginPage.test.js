import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { render, cleanup, fireEvent } from '@testing-library/react-native';
import { fromJS } from 'immutable';
import createSagaMiddleware from 'redux-saga';

import { setToken } from 'containers/App/actions';
import { logIn, logInFailure, logInSuccess } from '../actions';

import LoginPage from '..';


jest.mock('utils/request', () => ({
  request: (url, options) => {
    if(options.userPassword === "bad_password") throw Error("Could not log in.")
    return {
      token: 'token'
    }
  },
  getRequestOptions: (type, payload) => payload
}));

const initialState = {
  login: fromJS({
    errorMessage: '',
    loading: false,
  })
};

describe('<LoginPage />', () => {
  const sagaMiddleware = createSagaMiddleware();
  let store, rendered;

  const mockStore = (initialState) => ({ 
    ...configureStore([sagaMiddleware])(initialState), 
    injectedReducers: {}, 
    injectedSagas: {}, 
    runSaga: sagaMiddleware.run 
  });

  beforeEach( () => {
    store = mockStore(initialState);
    rendered = render(
      <Provider store={store}><LoginPage /></Provider>
    );
  })

  afterEach(cleanup);

  it('should render correctly', () => {
    const tree = rendered.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should display login form', () => {
    expect(rendered.getByPlaceholderText(/Email/i)).toBeTruthy();
    expect(rendered.getByPlaceholderText(/Password/i)).toBeTruthy();
    expect(rendered.getByText(/Log In/i)).toBeTruthy();
  });
  
  it('should disable button without input', () => {
    const button = rendered.getByText(/Log In/i);
    fireEvent(button, 'press');
    const actions = store.getActions();
    expect(actions).toHaveLength(0)
  });

  it('should dispatch login action, then loginSuccess', () => {
    fireEvent.changeText(rendered.getByPlaceholderText(/Email/i), "email@email.com");
    fireEvent.changeText(rendered.getByPlaceholderText(/Password/i), "password");
    fireEvent.press(rendered.getByText(/Log In/i));
    const actions = store.getActions();
    expect(actions[0]).toEqual(logIn("email@email.com", "password"));
    expect(actions[1]).toEqual(setToken('token'));
    expect(actions[2]).toEqual(logInSuccess());
  });

  it('should dispatch login action, then loginFailure', () => {
    fireEvent.changeText(rendered.getByPlaceholderText(/Email/i), "email@email.com");
    fireEvent.changeText(rendered.getByPlaceholderText(/Password/i), "bad_password");
    fireEvent.press(rendered.getByText(/Log In/i));
    const actions = store.getActions();
    expect(actions[0]).toEqual(logIn("email@email.com", "bad_password"));
    expect(actions[1]).toEqual(logInFailure("Could not log in."));
  });

  it('should render error message', () => {
    store = mockStore({ login: initialState.login.set('errorMessage', 'An error occurred') });
    rendered = render(
      <Provider store={store}><LoginPage /></Provider>
    );
    expect(rendered.getByText('An error occurred')).toBeTruthy();
  });
});