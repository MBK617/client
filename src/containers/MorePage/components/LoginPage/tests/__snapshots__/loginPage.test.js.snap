// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginPage /> should render correctly 1`] = `
<View
  style={
    Object {
      "height": "100%",
      "justifyContent": "center",
      "padding": 40,
    }
  }
>
  <View
    style={
      Object {
        "borderColor": "black",
        "borderWidth": 1,
        "marginBottom": 10,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Email"
      rejectResponderTermination={true}
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "borderColor": "black",
        "borderWidth": 1,
        "marginBottom": 10,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Password"
      rejectResponderTermination={true}
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <Text
    style={
      Object {
        "color": "#ff0000",
        "marginBottom": 10,
        "marginLeft": "auto",
        "marginRight": "auto",
        "textAlign": "center",
        "width": 250,
      }
    }
  >
    
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Log In
      </Text>
    </View>
  </View>
</View>
`;
